geth/account/accounts_test.go

```diff
@@ -103,9 +107,386 @@ func TestVerifyAccountPasswordWithAccountBeforeEIP55(t *testing.T) {
+	testErrWhisper    = errors.New("Can't return a whisper service")
+	testErrKeyStore   = errors.New("Can't return a key store")
+	testErrAccManager = errors.New("Can't return an account manager")
+)
+
+func TestManagerTestSuite(t *testing.T) {
+	nodeManager := newMockNodeManager(t)
+	accManager := NewManager(nodeManager)
+
+	keyStoreDir, err := ioutil.TempDir(os.TempDir(), "accounts")
+	require.NoError(t, err)
+	keyStore := keystore.NewKeyStore(keyStoreDir, keystore.LightScryptN, keystore.LightScryptP)
+	defer os.RemoveAll(keyStoreDir) //nolint: errcheck
+
+	testPassword := "test-password"
+
+	// Initial test - create test account
+	nodeManager.EXPECT().AccountKeyStore().Return(keyStore, nil)
+	addr, pubKey, mnemonic, err := accManager.CreateAccount(testPassword)
+	require.NoError(t, err)
+	require.NotEmpty(t, addr)
+	require.NotEmpty(t, pubKey)
+	require.NotEmpty(t, mnemonic)
+
+	s := &ManagerTestSuite{
+		testAccount: testAccount{
+			"test-password",
+			addr,
+			pubKey,
+			mnemonic,
+		},
+		nodeManager:    nodeManager,
+		accManager:     accManager,
+		keyStore:       keyStore,
+		shh:            whisper.New(nil),
+		gethAccManager: accounts.NewManager(),
+	}
+
+	suite.Run(t, s)
+}
+
+func newMockNodeManager(t *testing.T) *common.MockNodeManager {
+	ctrl := gomock.NewController(t)
+	return common.NewMockNodeManager(ctrl)
+}
+
+type ManagerTestSuite struct {
+	suite.Suite
+	testAccount
+	nodeManager    *common.MockNodeManager
+	accManager     *Manager
+	keyStore       *keystore.KeyStore
+	shh            *whisper.Whisper
+	gethAccManager *accounts.Manager
+}
+
+type testAccount struct {
+	password string
+	address  string
+	pubKey   string
+	mnemonic string
+}
+
+// reinitMock is for reassigning a new mock node manager to account manager.
+// Stating the amount of times for mock calls kills the flexibility for
+// development so this is a good workaround to use with EXPECT().AnyTimes()
+func (s *ManagerTestSuite) reinitMock() {
+	s.nodeManager = newMockNodeManager(s.T())
+	s.accManager.nodeManager = s.nodeManager
+}
+
+func (s *ManagerTestSuite) TestCreateAccount() {
+	s.reinitMock()
+
+	// Don't fail on empty password
+	s.nodeManager.EXPECT().AccountKeyStore().Return(s.keyStore, nil)
+	_, _, _, err := s.accManager.CreateAccount(s.password)
+	s.NoError(err)
+
+	s.nodeManager.EXPECT().AccountKeyStore().Return(nil, testErrKeyStore)
+	_, _, _, err = s.accManager.CreateAccount(s.password)
+	s.Equal(testErrKeyStore, err)
+}
+
+func (s *ManagerTestSuite) TestRecoverAccount() {
+	s.reinitMock()
+
+	s.nodeManager.EXPECT().AccountKeyStore().Return(s.keyStore, nil)
+	addr, pubKey, err := s.accManager.RecoverAccount(s.password, s.mnemonic)
+	s.NoError(err)
+	s.Equal(s.address, addr)
+	s.Equal(s.pubKey, pubKey)
+
+	s.nodeManager.EXPECT().AccountKeyStore().Return(nil, testErrKeyStore)
+	_, _, err = s.accManager.RecoverAccount(s.password, s.mnemonic)
+	s.Equal(testErrKeyStore, err)
+}
+
+func (s *ManagerTestSuite) TestSelectAccount() {
+	s.reinitMock()
```

Comment:
Shell we extract `s.reinitMock()` into `SetupTest`?

Reply:
Extracting? It's a separate function to use where necessary in order to drop all calls in `gomock.Controller`. Comment lines above `reinitMock` should make sense.

Reply:
Moving it into a function `SetupTest()` which automatically is executed for each test makes the manual calling in the beginning of each test obsolete. So not a direct extracting, but moving into a different method. That's what he meant.

Reply:
Alright then. That surely sounds like a better way to do it.